filebeat/input/filestream/internal/input-logfile/harvester_test.go

```diff
@@ -0,0 +1,286 @@
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		defer func() {
+			if v := recover(); v != nil {
+				t.Errorf("did not recover from harvester panic in defaultHarvesterGroup")
+			}
+		}()
+
+		goroutinesBeforeRun := runtime.NumGoroutine()
+
+		hg.Start(input.Context{Logger: logp.L(), Cancelation: context.Background()}, source)
+
+		for goroutinesBeforeRun < runtime.NumGoroutine() {
+			time.Sleep(1 * time.Millisecond)
+		}
+		assert.Nil(t, hg.StopGroup())
+	})
+
+	t.Run("assert a harvester error is handled", func(t *testing.T) {
+		mockHarvester := &mockHarvester{0, true}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+
+		goroutinesBeforeRun := runtime.NumGoroutine()
+
+		hg.Start(input.Context{Logger: logp.L(), Cancelation: context.Background()}, source)
+
+		for goroutinesBeforeRun < runtime.NumGoroutine() {
+			time.Sleep(1 * time.Millisecond)
+		}
+
+		_, ok := hg.readers.table[source.Name()]
+		assert.False(t, ok)
+
+		err := hg.StopGroup()
+		assert.Error(t, err)
+	})
+
+	t.Run("assert already locked resource has to wait", func(t *testing.T) {
+		mockHarvester := &mockHarvester{}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		inputCtx := input.Context{Logger: logp.L(), Cancelation: context.Background()}
+
+		goroutinesBeforeRun := runtime.NumGoroutine()
+		r, err := lock(inputCtx, hg.store, source.Name())
+		if err != nil {
+			t.Fatalf("cannot lock source")
+		}
+
+		hg.Start(inputCtx, source)
+
+		locked := true
+		for goroutinesBeforeRun < runtime.NumGoroutine() {
+			time.Sleep(10 * time.Millisecond)
+
+			if locked {
+				releaseResource(r)
+				locked = false
+			}
+		}
+
+		assert.Equal(t, 1, mockHarvester.runCount)
+		assert.Nil(t, hg.StopGroup())
+	})
+
+	t.Run("assert already locked resource has no problem when harvestergroup is cancelled", func(t *testing.T) {
+		mockHarvester := &mockHarvester{}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		inputCtx := input.Context{Logger: logp.L(), Cancelation: context.Background()}
+
+		goroutinesBeforeRun := runtime.NumGoroutine()
+		r, err := lock(inputCtx, hg.store, source.Name())
+		if err != nil {
+			t.Fatalf("cannot lock source")
+		}
+		defer releaseResource(r)
+
+		hg.Start(inputCtx, source)
+
+		for goroutinesBeforeRun < runtime.NumGoroutine() {
+			time.Sleep(100 * time.Millisecond)
+
+			err := hg.StopGroup()
+			assert.Error(t, err)
+		}
+
+		assert.Equal(t, 0, mockHarvester.runCount)
+	})
+}
+
+func testDefaultHarvesterGroup(t *testing.T, mockHarvester Harvester) *defaultHarvesterGroup {
+	return &defaultHarvesterGroup{
+		readers:   newReaderGroup(),
+		pipeline:  &pipelinemock.MockPipelineConnector{},
+		harvester: mockHarvester,
+		store:     testOpenStore(t, "test", nil),
+		tg:        unison.TaskGroup{},
+	}
+
+}
+
+type mockHarvester struct {
+	runCount  int
+	returnErr bool
+}
+
+func (m *mockHarvester) Run(ctx input.Context, s Source, c Cursor, p Publisher) error {
+	m.runCount += 1
+
+	timed.Wait(ctx.Cancelation, 1*time.Second)
```

Comment:
why we want to wait here? Maybe it should be optional and default to `<-ctx.Cancelation.Done()`.

By requiring the Run to block until cancellation we can be 'sure' that no resources have been leaked. The Wait and self-shutdown might hide issues.


e.g.

```
type mockHarvester struct {
	runCount  int
	returnErr bool
    do func(input.Context, Source, Cursor, Publisher) error
}

...

if m.do == nil {
  <- ctx.Cancelation.Done()
} else {
  err = m.do(ctx, s, c, p)
}
```

Now you can modify the behavior in a test if you want to

Reply:
I had to wait here so I can test that two harvesters cannot be started for the same source. But your proposal works much better. Thanks.